examples/isomorphisms.dx

```diff
@@ -0,0 +1,258 @@
+> (1, (2.0, 3))
+
+'Note that we assume but do not check that the isomorphism is lawful (i.e.
+`appIso iso $ revIso iso x == x` for all `x`, or equivalently
+`iso <>> (flipIso iso) == idIso`).
+
+'In addition, Dex will automatically write some useful isomorphisms for you
+to extract fields from records and variants. There are four syntactic forms
+that produce isos. We will start with the first two:
+- `#x` produces a "lens-like" record accessor `Iso {x:a & ...r} (a & {&...r})`
+- `#?x` produces a "prism-like" variant matcher `Iso {x:a | ...r} (a | {|...r})`
+
+%passes parse
+:t astype (Iso {a:Int & b:Float & c:Unit} _) #b
+> (Iso {a: Int32 & b: Float32 & c: Unit} (Float32 & {a: Int32 & c: Unit}))
+> === parse ===
+> _ans_:() =
+>   astype (Iso {a: Int & b: Float & c: Unit} _)
+>     (MkIso
+>        { bwd = \(x:(), r:()). {b = x, ...r}
+>        , fwd = \{b = x:(), ...r:()}. (,) x r})
+
+%passes parse
+:t astype (Iso {a:Int | b:Float | c:Unit} _) #?b
+> (Iso {a: Int32 | b: Float32 | c: Unit} (Float32 | {a: Int32 | c: Unit}))
+> === parse ===
+> _ans_:() =
+>   astype (Iso {a: Int | b: Float | c: Unit} _)
+>     (MkIso
+>        { bwd = \v:(). case v
+>                    ((Left x:())) -> {| b = x |}
+>                    ((Right r:())) -> {|b| ...r |}
+>                    
+>        , fwd = \v:(). case v
+>                    {| b = x:() |} -> (Left x)
+>                    {|b| ...r:() |} -> (Right r)
+>                    })
+
+'There are also two "zipper" forms, described later on this page:
+- `#&x` produces a "record-zipper" isomorphism
+  ```
+  Iso ({&...l} & {x:a & ...r}) ({x:a & ...l} & {&...r})
+  ```
+- `#|x` produces a "variant-zipper" isomorphism
+  ```
+  Iso ({|...l} | {x:a | ...r}) ({x:a | ...l} | {|...r})
+  ```
+
+'## Record accessors and lens-like helpers
+Record accessor isomorphisms can be passed into the helper function `getAt`:
+
+:t getAt
+> ((a:Type) ?-> (b:Type) ?-> (c:Type) ?-> (Iso a (b & c)) -> a -> b)
+
+:p getAt #foo {foo=1, bar=2.0}
+> 1
+
+'We can also do other types of things:
+
+:p popAt #foo {foo=1, bar=2.0}
+> {bar = 2.0}
+
+:p pushAt #foo 3.0 {foo=1, bar=2.0}
+> {bar = 2.0, foo = 3.0, foo = 1}
+
+:p setAt #foo 2 {foo=1, bar=2.0}
+> {bar = 2.0, foo = 2}
+
+'These helper functions work with any "lens-like" isomorphism. For instance,
+we can select everything except for a particular field:
+
+:t exceptLens
+> ((a:Type) ?-> (b:Type) ?-> (c:Type) ?-> (Iso a (b & c)) -> Iso a (c & b))
+
+:p getAt (exceptLens #foo) {foo=1, bar=2.0, baz=3}
+> {bar = 2.0, baz = 3}
+
+
+'## Variant accessors and prism-like helpers
+Similarly, there are prism-like helpers
+
+:t matchWith
+> ((a:Type) ?-> (b:Type) ?-> (c:Type) ?-> (Iso a (b | c)) -> a -> Maybe b)
+
+:t buildWith
+> ((a:Type) ?-> (b:Type) ?-> (c:Type) ?-> (Iso a (b | c)) -> b -> a)
+
+'which can be used with variant accessors or any other prism-like isomorphism:
+
+:p matchWith #?foo $ astype {foo:Int | bar:Float} {|foo = 1|}
```

Comment:
Couldn't this be just `{|foo = 1|} : {foo:Int | bar:Float}`? I'm not sure if `astype` helps that much with readability

Reply:
I don't think that actually works anymore, I think I broke that when #201 merged. The `variant : type` syntax has to be handled specially in the parser, and I removed that logic when I changed how variants were parsed.

I've wondered before, though, whether `expr : type` should be a primitive expression on its own that you can use anywhere? That might be useful, and would remove the need for `astype` (which is basically a poor man's `:`) One weird related thing is that I don't think there's a valid way of doing it in userspace, i.e. `(:)` should have type `a -> (a:Type) -> a` but that breaks scoping rules.

Reply:
Hah, yeah. The assertion is written out in the wrong order! I think we should just build it into a parser and we can have a primitive in the syntax language that corresponds to `astype`.

Reply:
Makes sense. I think I'll leave that for a future PR though and keep `astype` for now, if that's all right.

Reply:
Sure, sounds good.